### PR TITLE
Blob reader improvements

### PR DESCRIPTION
**Pull request by tmat**
[Work in progress]

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera  FYI

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs [hunk 1]
```diff
@@ -45,22 +47,37 @@ internal static MemoryBlock CreateChecked(byte* buffer, int length)
...
+
+            return newPointer;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private byte* GetPointerChecked(int offset, int byteCount)
         {
-            if (unchecked((ulong)(uint)offset + (uint)byteCount) > (ulong)Length)
+            byte* newPointer = StartPointer + offset;
+            if (newPointer < StartPointer || newPointer > EndPointer - byteCount)
```
Comment (jkotas) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs]: Can `EndPointer - byteCount` underflow?

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs [hunk 2]
```diff
@@ -13,14 +13,16 @@ namespace System.Reflection.Internal
...
-        internal readonly byte* Pointer;
-        internal readonly int Length;
+        internal readonly byte* StartPointer;
+        internal readonly byte* EndPointer;
+
+        internal int Length => (int)(EndPointer - StartPointer);
 
         internal MemoryBlock(byte* buffer, int length)
         {
             Debug.Assert(length >= 0 && (buffer != null || length == 0));
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs]: GetPointerChecked below does a check to see if StartPointer+offset would overflow and throws if it does.  Should we do the same here with buffer+length, even if just as an assert?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobReader.cs [hunk 3]
```diff
@@ -87,55 +86,43 @@ internal string GetDebuggerDisplay()
...
+                byte* newPointer = StartPointer + value;
 
-            _currentPointer = _block.Pointer + offset;
-            return true;
+                if (newPointer > _endPointer)
+                {
+                    Throw.OutOfBounds();
+                }
+
+                _currentPointer = newPointer;
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobReader.cs]: Could this just be:
```C#
_currentPointer = GetPointerChecked(value);
```
?

(Or, I may just be confused by the diff, and that GetPointerChecked is actually on a different type.)

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/IL/MethodBodyBlock.cs [hunk 4]
```diff
@@ -153,7 +153,7 @@ public static MethodBodyBlock Create(BlobReader reader)
             }
 
             var ilBlock = reader.GetMemoryBlockAt(0, ilSize);
-            reader.SkipBytes(ilSize);
+            reader.Offset += ilSize;
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/IL/MethodBodyBlock.cs]: Does set_Offset handle checking for overflows, or maybe that's not necessary here?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by karelz** [comment]
@tmat what are your plans to finish it off?

**Comment by tmat** [comment]
At some point I'll finish it off. Probably not for 1.1.0.

**Comment by karelz** [comment]
Can you please close the PR until it is ready for its prime time again? Thanks!

**Pull request closed without merging**